File: src/main/scala/com/barley/controllers/EmailController.scala
package com.barley.controllers

import java.lang.{Iterable, Long}
import java.util

import com.barley.model.{Attachment, Email, File}
import com.barley.respository.{AttachmentRepository, EmailRepository, FileRepository, RecipientRepository}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.{RequestMapping, RequestMethod, RequestParam, RestController}

/**
  * Created by vikram.gulia on 11/25/15.
  */
@RestController
@RequestMapping(value = Array("/emails"))
class EmailController @Autowired()(
                                    private val emailRepository: EmailRepository,
                                    private val attachmentRepository: AttachmentRepository,
                                    private val fileRepository: FileRepository,
                                    private val recipientRepository: RecipientRepository

                                  ) {

  val msg_query1 = "select message_id, outlook_id, subject, from_email, from_name, to_email, body_text, message_date, rank from " +
    "(select message_id, outlook_id, subject, from_email, from_name, to_email, body_text, message_date, @curRank := @curRank + 1 AS rank " +
    "from message, (SELECT @curRank := 0) r order by message_id) as messages where rank >"
  val msg_query2 = " and rank <= "

  @RequestMapping(value = Array("/list"), produces = Array("application/json"))
  def list(): Iterable[Email] = {
    emailRepository.findAll()
  }

  @RequestMapping(value = Array("/count"), method = Array(RequestMethod.GET))
  def count(): String = {
    "{ \"count\":" + emailRepository.count() + "}"
  }

  @RequestMapping(value = Array("/list"), method = Array(RequestMethod.GET), produces = Array("application/json"))
  def list(@RequestParam("page") page: Long, @RequestParam("number") numPerPage: Long): Iterable[Email] = {
    getEmails(page, numPerPage)
  }

  def getEmails(page: Long, numPerPage: Long): Iterable[Email] = {
    val ids: util.List[Long] = new util.ArrayList[Long]
    var start: Long = (page - 1) * numPerPage + 1
    val end: Long = ((page - 1) * numPerPage) + numPerPage
    while (end > start) {
      ids.add(start)
      start += 1
    }
    val emails: Iterable[Email] = emailRepository.findAll(ids)
    import scala.collection.JavaConversions._
    for (email <- emails) {
      val attachments: Iterable[Attachment] = attachmentRepository.findByMessageId(email.message_id)
      println(attachments)
      val fileIds: util.List[Long] = new util.ArrayList[Long]
      for (attachment <- attachments) {
        fileIds.add(attachment.file_id)
      }
      val fileIterable: Iterable[File] = fileRepository.findAll(fileIds)
      println(fileIterable)
      val files: util.List[File] = new util.ArrayList[File]
      for (file <- fileIterable) {
        files.add(file)
      }
      println(files)
      email.setFileArray(files)
    }
    emails
  }

}

File: src/main/java/com/barley/controllers/EmailController.java
package com.barley.controllers;

import com.barley.common.Utilities;
import com.barley.model.Email;
import com.barley.model.EmailAttach;
import com.barley.model.File;
import com.barley.model.FileAttach;
import com.barley.repository.AttachmentRepository;
import com.barley.repository.EmailRepository;
import com.barley.repository.FileRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.List;

@RestController
@RequestMapping(value = "/emails")
@CrossOrigin
public class EmailController {

    @Autowired
    private EmailRepository emailRepository;

    @Autowired
    private AttachmentRepository attachmentRepository;

    @Autowired
    private FileRepository fileRepository;

    @Autowired
    private Utilities utilities;

    @RequestMapping(value = "/list/count", method = RequestMethod.GET)
    public String count() {
        return "{ \"count\":" + emailRepository.count() + "}";
    }

    @RequestMapping(value = "/list/{message_id}", method = RequestMethod.DELETE)
    public void delete(@PathVariable(value = "message_id") Long message_id) {
        System.out.println("Received email to delete FIRST BLOCK " + message_id);
        emailRepository.delete(message_id);
        attachmentRepository.findByMessageId(message_id)
                .stream()
                .filter(att -> att != null)
                .forEach(attachment -> {
                            try {
                                if (attachment.getAttachment_type().equalsIgnoreCase("FILE")) {
                                    Long fileId = attachment.getFile_id();
                                    System.out.println("Delete Attachment id for file: " + fileId);
                                    if (fileId != null) {
                                        fileRepository.delete(fileId);
                                    }
                                } else {
                                    Long fileId = attachment.getMessageId();
                                    System.out.println("Delete Attachment id for email: " + fileId);
                                    if (fileId != null) {
                                        emailRepository.delete(fileId);
                                    }
                                }
                                attachmentRepository.delete(attachment.getAttachment_id());
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                );
    }

    @RequestMapping(value = "/list/{message_id}", method = RequestMethod.GET)
    public Email getOne(@PathVariable(value = "message_id") Long message_id) {
        return emailRepository.findOne(message_id);
    }

    @RequestMapping(value = "/list/{message_id}/html", method = RequestMethod.GET)
    public void getHtml(@PathVariable(value = "message_id") Long message_id, HttpServletResponse response) throws IOException {
        Email email = emailRepository.findOne(message_id);
        if (email != null && (email.getBody_html() != null || email.getBody_text() != null)) {
            response.setHeader("Content-Type", "text/html");
            response.setHeader("X-Frame-Options", "");
            byte[] bytes;
            String html = email.getBody_html();
            if (html != null && html.length() > 0)
                bytes = html.getBytes();
            else
                bytes = email.getBody_text().getBytes();
            utilities.prepareResponse(response, bytes);
        }
    }

    @RequestMapping(value = "/list", method = RequestMethod.GET, produces = "application/json")
    public List<Email> list(Pageable pageable) {
        return getEmails(pageable);
    }

    public List<Email> getEmails(Pageable pageable) {
        final Page<Email> emails = emailRepository.findAll(pageable);
        emails.forEach(email ->
                attachmentRepository
                        .findByMessageId(email.getMessage_id())
                        .stream()
                        .filter(att -> att != null)
                        .forEach(attachment -> {
                            if (attachment.getAttachment_type().equalsIgnoreCase("FILE")) {
                                Long fileId = attachment.getFile_id();
                                //System.out.println("Attachment Id: " + fileId);
                                if (fileId != null) {
                                    File file = fileRepository.findOne(attachment.getFile_id());
                                    if (file != null)
                                        email.getFileArray().add(new FileAttach(file.getLongFileName(), file.getFile_id()));
                                }
                            } else {
                                Long fileId = attachment.getFile_id();
                                //System.out.println("Attachment Id: " + fileId);
                                if (fileId != null) {
                                    Email em = emailRepository.findOne(fileId);
                                    if (em != null)
                                        email.getEmailArray().add(new EmailAttach(em.getMessage_id(), email.getSubject()));
                                }
                            }
                        })
        );
        return emails.getContent();
    }

}
